java-frontend/src/main/java/org/sonar/java/regex/ast/RegexSource.java

```diff
@@ -0,0 +1,106 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.regex.ast;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+
+/**
+ * The source code of a regular expression, made up out of 1 or more string literals
+ */
+public class RegexSource {
+
+  private final List<LiteralTree> stringLiterals;
+  private final String sourceText;
+  private final TreeMap<Integer, Integer> indices;
+
+  public RegexSource(List<LiteralTree> stringLiterals) {
+    this.stringLiterals = stringLiterals;
+    sourceText = stringLiterals.stream().map(RegexSource::getString).collect(Collectors.joining());
+    indices = new TreeMap<>();
+    int currentSourceIndex = 0;
+    int currentLiteralIndex = 0;
+    for (LiteralTree string : stringLiterals) {
+      indices.put(currentSourceIndex, currentLiteralIndex);
+      currentSourceIndex += string.value().length() - 2;
+      currentLiteralIndex++;
+    }
+  }
+
+  public Iterable<LiteralTree> getStringLiterals() {
+    return stringLiterals;
+  }
+
+  public String substringAt(IndexRange range) {
+    return sourceText.substring(range.getBeginningOffset(), range.getEndingOffset());
+  }
+
+  public String getSourceText() {
+    return sourceText;
+  }
+
+  public List<Location> locationsFor(IndexRange range) {
+    return locationsFor(range.getBeginningOffset(), range.getEndingOffset());
+  }
+
+  public List<Location> locationsFor(int beginningOffset, int endingOffset) {
+    List<Location> result = new ArrayList<>();
+    Position startPosition = findPosition(beginningOffset);
+    Position endPosition = findPosition(endingOffset);
+    for (int i = startPosition.indexOfLiteral; i <= endPosition.indexOfLiteral; i++) {
+      LiteralTree literal = stringLiterals.get(i);
+      int length = literal.value().length() - 2;
+      int startIndex = startPosition.indexOfLiteral == i ? startPosition.indexInsideLiteral : 0;
+      int endIndex = endPosition.indexOfLiteral == i ? endPosition.indexInsideLiteral : length;
+      result.add(new Location(literal, startIndex, endIndex));
+    }
+    return result;
+  }
+
+  private Position findPosition(int sourceIndex) {
+    Map.Entry<Integer, Integer> entry = indices.floorEntry(sourceIndex);
+    return new Position(entry.getKey(), entry.getValue());
```

Comment:
What about a `Position` constructor which exactly this instead of a dedicated method? Of course then the class can not be `static` anymore, but I think it makes sense.

```
Position startPosition = new Position(beginningOffset);
Position endPosition = new Position(endingOffset);
```
...
```
private class Position {
  Position(int sourceIndex) {
    Map.Entry<Integer, Integer> entry = indices.floorEntry(sourceIndex);
    this.indexOfLiteral = entry.getKey();
    this.indexInsideLiteral = entry.getValue();
  }

...
```

Reply:
That makes sense, I'll do that. Thank you.